jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferOutputStream2.java

```diff
@@ -0,0 +1,169 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+import org.eclipse.jetty.util.BufferUtil;
+
+/**
+ * This class implements an output stream in which the data is written into a list of ByteBuffer,
+ * the buffer list automatically grows as data is written to it, the buffers are taken from the
+ * supplied {@link ByteBufferPool} or freshly allocated if one is not supplied.
+ *
+ * Designed to mimic {@link java.io.ByteArrayOutputStream} but with better memory usage, and less copying.
+ */
+public class ByteBufferOutputStream2 extends OutputStream
+{
+    private static final int DEFAULT_BUFFER_SIZE = 1024;
+
+    private final boolean _direct;
+    private final ByteBufferPool _bufferPool;
+    private final List<ByteBuffer> _buffers = new LinkedList<>();
+    private final int _bufferSize;
+    private int _size = 0;
+
+    public ByteBufferOutputStream2()
+    {
+        this(null);
+    }
+
+    public ByteBufferOutputStream2(ByteBufferPool bufferPool)
+    {
+        this(bufferPool, true, DEFAULT_BUFFER_SIZE);
+    }
+
+    public ByteBufferOutputStream2(ByteBufferPool bufferPool, boolean direct, int bufferSize)
+    {
+        _bufferPool = (bufferPool == null) ? new NullByteBufferPool() : bufferPool;
+        _direct = direct;
+        _bufferSize = bufferSize;
+    }
+
+    /**
+     * Get an aggregated content written to the OutputStream in a ByteBuffer.
```

Comment:
The javadoc needs to make it REALLY clear that this buffer has been taken from the pool and needs to eventually be returned

Reply:
The buffer does not need to be returned to the pool by the user, currently it will be returned when the `OutputStream` is closed. See line 86 where we add the combined buffer back to the list replacing the other ones.

Reply:
then that needs to be made really clear!!!
I'm getting very luke warm about this as if we screw up the buffer pool we will have big problems.